internal/devserver/main/main.go

```diff
@@ -32,7 +40,71 @@ func main() {
 		os.Exit(1)
 	}
 
-	http.Handle(*servingPath, concatjs.ServeConcatenatedJS(*manifest, *base, nil /* realFileSystem */))
+	concatFilesManifestPath := filepath.Join(*base, *concat_files_manifest)
+	concat_files, err := manifestFiles(concatFilesManifestPath)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to read concat_files_manifest: %v\n", err)
+		os.Exit(1)
+	}
+
+	nodeModulesManifestPath := filepath.Join(*base, *node_modules_manifest)
+	node_modules_files, err := manifestFiles(nodeModulesManifestPath)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to read node_modules_manifest: %v\n", err)
+		os.Exit(1)
+	}
+
+	scripts := make([]string, 0, 100)
+
+	requireJsPath := resolveNodeModulesFile(node_modules_files, "/node_modules/requirejs/require.js");
```

Comment:
can this go program be unaware of requirejs, and just do that as the default value for the concat_files attribute on the rule?

Reply:
For requirejs since devserver expects its version from package.json, I search the rules_typescript node_modules first and then the user node_modules. I don't do this for user supplied scripts in concat_files as it wouldn't make sense. So there is a difference.

Reply:
It doesn't make sense because there is only one node_modules to resolve to, right?

Reply:
No resolution for the concat_files. I opted for labels so if the user wants to use a node_modules files as a script then they would just point the exact file they want so there is no ambiguity and the build would fail if the file is not there.

Reply:
But yes, if concat_files was resolved then it would be just the user @//:node_modules to look through and not the rules_typescript node_modules first.

Reply:
We can also supply labels for the requirejs dependency. If we specified two labels for that, we could have all the lookup locations we need. 
```
concat_scripts = ['$(rootpath @build_bazel_rules_typescript//:node_modules/requirejs/require.js)', '$(rootpath @//:node_modules/requirejs/require.js'] + concat_scripts
```
Then I guess the only thing would be to avoid erroring if the first one of those doesn't exist.

I keep thinking this is too complex. I think we should have rules_typescript install its own node_modules and stop relying on a layout that might be mixed with the users. https://github.com/bazelbuild/rules_nodejs/pull/67 should be a way for rules_typescript to guarantee it has a `@rules_typescript_node_modules//:all/the/files/we/expect.js`

Reply:
That could work for require.js. The go binary can skip files that are missing without a warning as bazel would have checked that the user supplied files exist.

I agree about having rules_typescript handle its own node_modules. You guaranteed to get what you expect where you expect it and you could also add @build_bazel_rules_typescript at a git repository instead of a local one since you're not relying on the user's npm to install the dependencies.

Reply:
Latest thinking on this:
bazel installing `git_repository(name = "@build_bazel_rules_typescript")` means you get a source dependency on tsc_wrapped

npm installing `@bazel/typescript` should probably be a compiled distro with .js/.d.ts files, and not require the user to compile tsc_wrapped, which would cut our dependencies to not include our compile-time dependencies, like our standard best practice for npm distros

We could possibly support both, even though we know the user is using Bazel so we could always rely on source distribution.

In the latter case, is there really a need for `ts_repositories()` to install our transitive deps?

If we do have a ts_repositories, it would be nice to have `npm_install(name = "build_bazel_rules_typescript_devserver_runtime_deps")` so that users don't download these packages unless they run a `ts_devserver` target. So we would have a separate `package.json` file with the `require.js` and `livereload` packages in `devDependencies`.

In other words, we have the option to avoid mixing dependencies between `ts_library` and `ts_devserver` which seems like the idiomatic Bazel thing to do.

Let's figure this out today and get it merged.

Reply:
Makes sense to me